Revamp bounds declaration checking rules for expressions with control flow.

dtarditi opened the pull request:
C only partially defines the order of evaluation of expressions.  It also allows expressions to have nested assignments and control flow.  This specification change revamps the description of checking of bounds declarations to account for these language features properly.   The specification before handled undefined order of evaluation well, but did not handle sequential control-flow (the comma operator) or conditional control-flow (conditional expressions and logical Boolean operators) well.

We specify checking using mutually-recursive functions that take C expressions and additional information about program state, and produce new information about program state after the expressions are evaluated.  The information that is tracked includes contexts that map variables to bounds expressions, sets of sets of equivalent expressions, and pending variables (variables that have unordered assignments with respect to the current expression being checked).   The functions may also produce an error if the meaning of the expression is undefined according to the C specification because of unordered assignments or uses of variables that are unordered with respect to assignments.

This change also adds a concept of temporaries introduced/required by checking so that we can more easily describe bounds that use the value produced by an expression, as well as bounds of expressions with nested side-effects.  It removes current_expr_value.

This change describes some things that are already being done in the Checked C clang implementation (the use of temporaries and sets of sets of equivalent expressions).   It also provides a design that can be used to check expressions with control flow.  This design will replace the current implementation effort to use the clang CFG for checking expressions with control flow.  There were two problems with the approach that used the clang CFG.  First, it was difficult to explain at the source program level what was happening.  Second, the clang CFG overspecifies the order of evaluation of expressions.  The checking would not have signaled an error when the meaning of the expression was actually undefined.  The clang CFG expression evaluation order isn't guaranteed to be used by the actual machine code.  The order of evaluation could differ. This could potentially allow subtle order-of-evaluation issues to compromise the integrity of bounds.  The new design projects against these kinds of issues.

lenary commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+x = alloc(5), len = 5;
+\end{lstlisting}
+
+Here is a list of checking steps:
```
Nice, this was really helpful.

lenary commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+  \>\>\>$CheckLValue($\code{x}$, C = \{x \mapsto$\code{bounds(x, x + len)}$\}, EQ = \emptyset, PV = \emptyset)$ \\
+  \>\>\>\>$(B = $ \code{bounds(&x, &x + 1)}, $RB = $\code{bounds(x, x + len)}$,
+                           UC = \{x \mapsto$\code{bounds(x, x + len)}$\},$\\
+  \>\>\>\>~$UEQ = \emptyset, G = \{$\code{&v}\}$)$ \\
```
I don't understand where the `G={&v}` came from.

dtarditi replied:
This should be &x.  I'll fix it in my next push to the PR branch.

lenary commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+regarding propagating bounds from the prior case can be applied to show that the order of
+assignments does not matter.
+\end{itemize}
+\end{itemize}
```
This proof needs a paragraph to wrap it up - something like "Similar reasoning applies where `Check` is called before `CheckLValue`. Thus we have shown..."

lenary commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+Suppose that $lhs$
+contains a single assignment expression that updates a variable $v$ with an expression $e1$ and 
+$rhs$ contains a single assignment expression that updates  a variable $w$ with an expression $e2$.  
+We know that $lhs$ does not update or use $w$ because of the pending variable parameters.  Similarly, 
```
I think these two sentences would be more clearly stated as "We assume that `w` does not appear in `e1`, and `v` does not appear in `e2`." from which the set of pending variables should be more obvious.

lenary commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -1515,93 +2332,15 @@ \section{Other statements}
+before the evaluation of the full expression. No way is provided for a
+full expression occurring in a statement to have new bounds declared 
+for any variables in it. This means that the bounds declarations in the
+context will be expected to be true after the evaluation of the full expression too.
```
This means I can't do, for instance:
```
for (array_ptr<int> p = q : count(len); p < q + len; p++) {
  ...
}
```
Ok, that seems reasonable, and something we can work to relax in the future.

lenary replied:
Let's turn this into a follow-up issue.

lenary commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -1515,93 +2332,15 @@ \section{Other statements}
+context will be expected to be true after the evaluation of the full expression too.
 
-The rules in Section~\ref{section:checking-nested-assignment-expressions}
+The rules in Section~\ref{section:checking-compleex-expressions}
```
Nit: typo
```suggestion
The rules in Section~\ref{section:checking-complex-expressions}
```

dtarditi replied:
Fixed.

lenary commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -83,34 +137,29 @@ \section{Preparing a translation unit for checking}
+We first discuss inferring bounds expressions for expressions. The bounds for an
+expression are always determined with respect to a bounds
+declaration context, which maps variables to their
+expressions.  We will use $\vdash$ to denote the
```
Nit: clarity
```suggestion
bounds expressions.  We will use $\vdash$ to denote the
```

dtarditi replied:
Fixed.

lenary commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -157,6 +206,16 @@ \subsection{Variables}
+
+During bounds checking, a temporary is introduced for the result
+of evaluating a string literal or compound literal expression.
+If a literal \var{e} has array type with \var{n} elements, then
```
Nit: IIRC 'n' does not include the null character at the end of strings. This should be noted.

dtarditi replied:
Fixed.   This made the description longer, so it is worth looking at the new description.

lenary commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -83,34 +137,29 @@ \section{Preparing a translation unit for checking}
+may happen at some less-precisely specified time.   The evaluation of
+expressions at runtime needs to store values in a temporary, 
+if the temporary is used in a runtime
+bounds check.
```
At some point we need to clarify the scope of these temporaries. Something to think about for the moment, and insert later. My main worry would be these temporaries "leaking" out of scope. Potentially that leakage can be avoided using the sets of equivalent expressions.

dtarditi replied:
The temporaries will need to have function scope and lifetime.  Currently, I believe they will be scoped to an expression statement or a declarator with an initializer.    For checking of bundled scopes, they will need to propagate beyond their current scopes.  I've opened issue #345 to track adding this information to the specification.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -1,45 +1,99 @@
+
+To handle both assignments to pointer variables with bounds and assignments
+to variables used in bounds, we generalize bounds declaration
+checking to operate over contexts.  Contexts are maps from variables
```
Fixed.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -1,45 +1,99 @@
+
+To handle this, the analysis tracks variables where assignments
+to those variables may not have been completed yet.  We call these
+variables pending variables.  The analysis models indeterminate order
```
Updated.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -1,45 +1,99 @@
+to those variables may not have been completed yet.  We call these
+variables pending variables.  The analysis models indeterminate order
+of evaluation by placing variables in pending sets, and sequential
+control-flow by not placing variables in pending sets.  The analysis
```
I removed the sentence about nothing happening with a defined  order of evaluation for pending sets.   I think it is confusing to talk about that here.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -157,6 +206,16 @@ \subsection{Variables}
+\subsection{String literals and compound literal expressions}
+A string literal \lstinline+''abc''+ evaluates to an lvalue with array type,
+as does a compound literal of the form \lstinline+(int []) { 0, 1, 2, 3}+.
+Thes lvalues are then converted implicitly to values with pointer type.
```
Fixed.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -298,32 +359,38 @@ \subsection{Function calls}
 
 Next, the actual arguments are substituted for the formal parameters and
-\exprcurrentvalue\ is substituted for \lstinline|return_value|.
+a temporary \var{t1} is substituted for \lstinline|return_value|.
```
I added the clarifying modifier "new", as you suggested.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -1,45 +1,99 @@
+to variables used in bounds, we generalize bounds declaration
+checking to operate over contexts.  Contexts are maps from variables
+to bounds expressions for the values stored in the variables.  For
+each full expression in a function, we start with a context based
```
As you noted, full expression wasn't defined until much later in the section.  I've moved the definition earlier and italicised the first use of the term.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -548,30 +517,71 @@ \subsection{Comma expressions}
+We use the updated context and sets of equivalent expressions to check
+that the new bounds expressions subsume the declared bounds expressions for variables.
 
 For an assignment expression of the form \var{x} \code{=} \var{e},
```
Yes, this was confusing to follow.  I've grouped the 3 item together, and then discuss the 2nd item in detail.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -548,30 +517,71 @@ \subsection{Comma expressions}
+This section describes how to check bounds declarations for
+simple assignments to variables.  Thiese are expression statements of the
+form \code{v = e1}, where \code{v} is a variable and \code{e1} does
+not contain nested assignments.
```
I added a forward reference.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -621,19 +631,12 @@ \section{Bounds for assignment expressions}
-expression whose bounds is being determined. Because the value of
-\var{e} has been assigned to \var{x}, \var{x} is substituted for
-\exprcurrentvalue\ in the bounds for x.
+Third, if the value of \var{e} is also available in a temporary,
```
This has been fixed.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+variables are valid.
+\item $CheckFullExpr$: checks that declared bounds are valid after
+evaluating a full expression.  
+A full expression \cite[Annex C]{ISO2011} is an expression not nested within another expression.
```
Yes, I moved this earlier.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+
+The function $Pending(e)$ uses the helper functions,
+$GetLValueVariable(e)$ and $GetRValueVariable(e)$.  Those functions
+return a set containing a variable name or the empty set.
```
I added the clarifying term as you suggest.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+
+The checking is done by the following functions:
+\begin{enumerate}
+\item $Pending(e)$: calculates the set of variables $PV$
```
I changed the descriptions to not declare arguments here.  I started on the the use of mathit, but decided that is better done as a separate change.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+\item  Otherwise, let $B_x^\prime = B_x$ with $OV$ substituted for the use of $v$.
+Let $UC = UC[x \mapsto B_x^\prime]$
+\end{enumerate}
+\item Let $UB = B$.  If $UB$ uses $v$, update $UB$ similarly.
```
I expanded the description to avoid any confusion.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+Let $UC = UC[x \mapsto B_x^\prime]$
+\end{enumerate}
+\item Let $UB = B$.  If $UB$ uses $v$, update $UB$ similarly.
+\item Record the new bounds for $v$: let $UC = UC[v \mapsto UB]$.
```
I'm not too worried about efficiency  at this level of description.  You are correct that we'd update the map twice.  It seemed more complex to add a clause excluding the entry for $v$ from the first update.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+\end{enumerate}
+\item Let $UB = B$.  If $UB$ uses $v$, update $UB$ similarly.
+\item Record the new bounds for $v$: let $UC = UC[v \mapsto UB]$.
+\item Let $UEQ = EQ$.  Adjust $UEQ$ so that it is true for the program state after the assignment. For any expression $d$ in $UEQ$ that uses the value of $v$, if $OV$ is {\it None}, let $UEQ$ = $UEQ $ - $d$.
```
Yes, it is the latter.  I've clarified the formatting to make this clear.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+Check(rhs, {UC}_{lhs}, {UEQ}_{lhs},P_{rhs}, Form, RB)$.
+\item If $NeedsBoundsCheck(lhs)$ returns true, check that $B_{lhs}$ does not use any variables in 
+$PV \cup Pending(lhs) \cup Pending(rhs)$.  If it does, report an error.
+\item  If $lhs$ is a variable $v$ (that is, $GetLValue(lhs) = { v }$),
```
Fixed - thanks for catching this!

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+      as $Val$ but is a non-modifying expression. 
+      
+      For any subexpression $S_i$ that is a modifying expression, 
+      use an expression from $G_i$ (or $G_1$ when $Form = PostIncDec$). If $G_i$ is empty, $Val\prime$ cannot
```
Fixed.

dtarditi commented:
Thank you for the thorough review!  I addressed most of your points.  There are still a few points to address.  I plan to work on them tomorrow.

Also, thank you or the suggestion to use \mathit.  I noticed that the spacing was wrong, but wasn't sure how to fix it.  Using mathot will introduce a lot of formatting changes.  I started on that and decided I should fix other issues first.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -548,30 +517,71 @@ \subsection{Comma expressions}
+\label{section:checking-simple-assignment}
+
+This section describes how to check bounds declarations for
+simple assignments to variables.  Thiese are expression statements of the
```
Fixed.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+\item Let $RB = $ \boundsunknown.
+\end{enumerate}
+\end{enumerate}
+\item If $e$ has the form $*e1$,  let $(B, UC, EQ, \_) = Check(e1, C, EQ, PV, No)$.  Let $G =\emptyset$ and $RB = $ \boundsunknown.
```
I've added clarifying information for these cases..  Checked C doesn't allow bounds to be declared for individual array elements  or values pointed to by pointers.

We do use types when we can to construct bounds for these cases.  For example, `Ptr<T>` implies bounds of a single element of type T.  I've opened issue #344 to track using types to construct bounds.

dtarditi commented:
@lenary thanks for the detailed feedback.   I think I have addressed almost all of your feedback.  This is ready for re-review.  I opened issues #343 and #344 to track remaining work (fixing formatting and taking advantage of type information when computing the bounds for values stored in a lvalue).

lenary commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -1,45 +1,101 @@
+expressions via comma expressions, conditional expressions, and
+logical boolean expressions (the \code{&&} and \code{||} expressions).
+
+To handle this, the analysis tracks sets variables where assignments
```
Nit: missing word
```suggestion
To handle this, the analysis tracks sets of variables where assignments
```

lenary commented:
Ok, re-reviewing a big review like this isn't very much fun on Github. 

This is almost entirely because GitHub cannot be bothered to show all the review comments (When loading the page, I get "27 hidden conversations", which when clicked, leaves me with "7 hidden conversations", and only after clicking that is everything shown). I think there are still unaddressed comments (and Nits) in those hidden conversations.

The other problem GitHub has is it cannot keep track of comments as the diffs become outdated, which makes working out if they're resolved or not a bit harder. 

In those comments, the important ones are (so you know you've addressed them):
- [ ] The comment about the description of Validate perhaps not being enough.
- [ ] Where does `G={&v}` come from?
- [ ] Setting up pending variables (relating to `v` and `w`) in the proof of different orderings of recursive checking.
- [ ] Concluding that proof properly.

I'm going to hit "resolve" on all the comments that were entirely compliments, so it should be clearer what's still unresolved.

I didn't do this as a real review because i wasn't sure if it would blow away the unresolved comments from the prior review.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+because an expression with an undefined order of evaluation of side-effects might have an 
+undefined meaning.  Checking must produce an error for these expressions because the
+validity of bounds information is not maintained by such expressions. In addition, checking must still
+model the order of side-effects when the order is defined.
```
I added a paragraph describing this. (it is about 3 paragraphs below this note)  You might want to take a look at it.

dtarditi commented on spec/bounds_safety/checking-variable-bounds.tex:
```diff
@@ -781,46 +785,842 @@ \subsection{Invertible expressions}
+\begin{itemize}
+\item  If both arms produce the same updated context, we use that context
+and merge the rest of the results.
+\item Iif the arms produce different
```
This will be fixed in next push to the PR branch.

dtarditi commented:
OK, I think I've addressed the remaining comments that I missed initially.   I opened issue #348 as a follow-up issue for flow-sensitive bounds declarations within an expression.  

Thanks again @lenary for reviewing this!  I know GitHub's UI is cumbersome for large reviews.
